### DISCUSSION TITLE
Couldn't get GraphQL subscriptions to work

### DISCUSSION THREAD

**Discussion by zbhavyai** [quarkusio/quarkus#25024] [Q&A] [answered]
Hello,

I am following the quickstart given at [https://quarkus.io/guides/smallrye-graphql](https://quarkus.io/guides/smallrye-graphql), and was trying to get the GraphQL subscriptions to work. I understand from the note that _Subscription is currently still considered experimental_.

I followed the quickstart guide exactly, and made some changes to the `FilmResource` class by adding the subscription code given in the guide. Modified `FilmResource` can be found here [https://gist.github.com/zbhavyai/ab1b0818524270a53177743d762e419d](https://gist.github.com/zbhavyai/ab1b0818524270a53177743d762e419d). The modifications are done in a local copy of [https://github.com/quarkusio/quarkus-quickstarts/tree/main/microprofile-graphql-quickstart](https://github.com/quarkusio/quarkus-quickstarts/tree/main/microprofile-graphql-quickstart).

Upon making ws connection with graphql-ui, it prints the message `"Connection initialized (protocol=graphql-transport-ws), requested a subscription..."`, but doesn't output any added `Hero` details (screenshot attached below). I tried this with Quarkus versions `2.8.0.Final` and `2.8.1.Final`.

I came across the issue #21434, and was wondering if this concerns the problem I am facing.

So, I would really like to know if I am doing it the right way, and if subscriptions really work?

![subscription_not_working](https://user-images.githubusercontent.com/14868531/164146546-b804d250-270a-47b5-b91b-57636c7cc7e4.png)

Thank you.

[a comment by an automated account is not shown]

**Comment by phillip-kruger** [comment]
Hi @zbhavyai - do you have a small reproducer I can look at ? Thanks

**Reply by zbhavyai** [reply]
Hi @phillip-kruger 
Here is code I tried - [https://github.com/zbhavyai/quarkus-subscription-code-share](https://github.com/zbhavyai/quarkus-subscription-code-share).

Thanks

**Comment by phillip-kruger** [comment, marked as answer]
Hi @zbhavyai - Thanks. Have you tried with older version of Quarkus ? (Just to test). This should work, but it does not. @jmartisk can you have a look too ?

**Reply by zbhavyai** [reply]
@phillip-kruger No, I haven't tried it with any other version of Quarkus, other than the `2.8.1.Final` and `2.8.0.Final`, but can try it. Which older version of Quarkus would you suggest?

**Reply by phillip-kruger** [reply]
Not sure, try 2.7.0, 2.6.0, 2.5.0 ....

**Reply by zbhavyai** [reply]
I tried the code with Quarkus versions `2.7.0.Final`, `2.6.0.Final`, `2.5.0.Final`, and `2.4.0.Final`. GraphQL subscriptions doesn't seem to work with any of these. In case you want to check the `pom.xml`/code, it is in the [same](https://github.com/zbhavyai/quarkus-subscription-code-share) repo, with branches named according to `test-{version}`.

**Reply by phillip-kruger** [reply]
Ok that is very weird as it was working and we have tests for this. @jmartisk please can you have a look ?

**Reply by jmartisk** [reply]
I'll look into it in a bit

**Reply by jmartisk** [reply]
The 'connection initialized' message means that you have a subscription going, but there haven't been any items yet. The question is why you're not getting any items.
Not completely sure but it seems to be something with the usage of Mutiny.
You seem to be using a BroadcastProcessor (https://github.com/zbhavyai/quarkus-subscription-code-share/blob/043fe291a0d168d2ede73bf1a0287bd9d609f5fa/src/main/java/org/acme/microprofile/graphql/FilmResource.java#L23) which I haven't used and I don't know how it works, but apparently if you return it as a Multi, then that Multi doesn't receive events from calling `onNext`. If I change the return statement in the `@Subscription` method to something like `Multi.createFrom().item(new Hero());` then the client receives it...

**Reply by zbhavyai** [reply]
@jmartisk Unfortunately, the suggested alternative statement doesn't work as intended. If I use what you suggested, it will print the details only once, and the ws connection seems to end after printing; it's not really listening to future events. What I want is the ws connection to remain open, and data (which is `Hero` details in our case) should keep printing when new `Hero` is added.

I had used `BroadcastProcessor` as it was mentioned in the [quickstart](https://quarkus.io/guides/smallrye-graphql#subscriptions) guide.

So, do you think this is a bug?

**Reply by jmartisk** [reply]
So, I dug deeper into it and now I know what the problem is!
The usage of the `BroadcastProcessor` itself is correct, but the problem is that your `@GraphQLApi` endpoint is request-scoped. That means that each time you invoke a mutation, it will execute in a new instance of the `FilmResource` class, and therefore with a different instance of the `BroadcastProcessor` than what was originally returned from the subscription method, therefore the `onNext` call to the processor won't be properly matched to the running subscription. You need to annotate the `FilmResource` with `@ApplicationScoped` for this to work.
Let me know if this fixes your issue.
I think I will submit a PR to add `@ApplicationScoped` to the `FilmResource` class in the quickstart to prevent this confusion from happening again in the future.

**Reply by phillip-kruger** [reply]
Good catch ! Nice @jmartisk ! Thanks

**Reply by jmartisk** [reply]
I've submitted a PR that incorporates the whole subscription-related part of the guide into the quickstart: https://github.com/quarkusio/quarkus-quickstarts/pull/1105

**Reply by zbhavyai** [reply]
@jmartisk Many thanks! Now it works.
